lib/auth/crownjewel/crownjewelv1/service_test.go

```diff
@@ -0,0 +1,221 @@
+			t.Run("allowed admin states", func(t *testing.T) {
+				for _, state := range tt.allowedStates {
+					t.Run(stateToString(state), func(t *testing.T) {
+						for _, verbs := range utils.Combinations(tt.allowedVerbs) {
+							t.Run(fmt.Sprintf("verbs=%v", verbs), func(t *testing.T) {
+								service := newService(t, state, fakeChecker{allowedVerbs: verbs})
+								err := callMethod(t, service, tt.name)
+								// expect access denied except with full set of verbs.
+								if len(verbs) == len(tt.allowedVerbs) {
+									require.False(t, trace.IsAccessDenied(err))
+								} else {
+									require.True(t, trace.IsAccessDenied(err), "expected access denied for verbs %v, got err=%v", verbs, err)
+								}
+							})
+						}
+					})
+				}
+			})
+
+			// test the method with disallowed admin states; expect failures.
+			t.Run("disallowed admin states", func(t *testing.T) {
+				disallowedStates := otherAdminStates(tt.allowedStates)
+				for _, state := range disallowedStates {
+					t.Run(stateToString(state), func(t *testing.T) {
+						// it is enough to test against tt.allowedVerbs,
+						// this is the only different data point compared to the test cases above.
+						service := newService(t, state, fakeChecker{allowedVerbs: tt.allowedVerbs})
+						err := callMethod(t, service, tt.name)
+						require.True(t, trace.IsAccessDenied(err))
+					})
+				}
+			})
+		})
+	}
+
+	// verify that all declared methods have matching test cases
+	t.Run("verify coverage", func(t *testing.T) {
+		for _, method := range crownjewelv1.CrownJewelService_ServiceDesc.Methods {
+			t.Run(method.MethodName, func(t *testing.T) {
+				match := false
+				for _, testCase := range testCases {
+					match = match || testCase.name == method.MethodName
+				}
+				require.True(t, match, "method %v without coverage, no matching tests", method.MethodName)
+			})
+		}
+	})
+}
+
+var allAdminStates = map[authz.AdminActionAuthState]string{
+	authz.AdminActionAuthUnauthorized:         "Unauthorized",
+	authz.AdminActionAuthNotRequired:          "NotRequired",
+	authz.AdminActionAuthMFAVerified:          "MFAVerified",
+	authz.AdminActionAuthMFAVerifiedWithReuse: "MFAVerifiedWithReuse",
+}
+
+func stateToString(state authz.AdminActionAuthState) string {
+	str, ok := allAdminStates[state]
+	if !ok {
+		return fmt.Sprintf("unknown(%v)", state)
+	}
+	return str
+}
+
+// otherAdminStates returns all admin states except for those passed in
+func otherAdminStates(states []authz.AdminActionAuthState) []authz.AdminActionAuthState {
+	var out []authz.AdminActionAuthState
+	for state := range allAdminStates {
+		found := slices.Index(states, state) != -1
+		if !found {
+			out = append(out, state)
+		}
+	}
+	return out
+}
+
+// callMethod calls a method with given name in the DatabaseObjectService service
+func callMethod(t *testing.T, service *Service, method string) error {
+	for _, desc := range crownjewelv1.CrownJewelService_ServiceDesc.Methods {
```

Comment:
The doc for `ServiceName_ServiceDesc` states

```
// It's only intended for direct use with grpc.RegisterService,
// and not to be introspected or modified (even as a copy)
```

but I can't find a useful way to make use of `protoreflect`/`protoregistry`, and I'm not entirely sure if going through `reflect` is better than going through an exported var that Must Not Be Used (and that other projects are indubitably also using).

<details>
<summary>for posterity</summary>

```go
	m := reflect.ValueOf(service).MethodByName(method)
	if !m.IsValid() {
		require.FailNow(t, "method %v not found", method)
	}
	r := m.Call([]reflect.Value{
		reflect.ValueOf(context.Background()),
		reflect.New(m.Type().In(1).Elem()),
	})
	err := r[1].Interface()
	if err == nil {
		return nil
	}
	return err.(error)
```

</details>

Reply:
>and not to be introspected or modified

Reflect sounds like introspecting 🤔 I'll leave the current version as it looks "safer" IMO